### DISCUSSION TITLE
C Function API

### DISCUSSION THREAD

**Discussion by danz68** [duckdb/duckdb#8628] [Ideas]
Recently the Python Function API was made available for developers, which gives a great flexibility to data science users. Is it possible to expose a similar interface to C developers (I am talking strictly about a dedicated C Function API, which is clearly different from C++ API)

Providing interfaces for Scalar UDF (params scalar, ret scalar)  as well as Vectorial UDF (params vector, ret vector) and Aggregate UDF (params vector, ret scalar) would be an useful addition for developers. For vectorial params/ret using native DataChunks would be ideal. For those who have already Arrow manipulation libraries an exposed DataChunk2Arrow and Arrow2DataChunk would be handy as well. 

Sure, because one DataChunk doesnt necessarily contain all required data for a vectorial or aggregate computation, I assume that streaming DataChunk structure for both vector reads and writes would be ideal, leaving to the developer the responsibility of streaming, locally storing and scheduling the streaming packages (can write back chunk results after each chunk read, or  read all first, store local, process and write back chunk results). For performance, a thread safe approach (reading/writing  chunks independently) would be great. 

Dreaming too much?

**Comment by Mause** [comment]
This has already been available for some time: https://duckdb.org/docs/archive/0.8.1/api/c/table_functions

**Reply by Tishj** [reply]
Those are table functions, not scalar functions.
We have support for scalar UDFs in the c++ API, it could be exposed to C, happy to review a PR @danz68

**Comment by danz68** [comment]
I probably need to describe better the vector UDF functionality I have in mind.

Let me use a LOESS like smoothing function for example. A common LOESS function goal is to compute a smooth curve which approximates an original set/vector of X,Y pairs, using a constant parameter as the smoothing factor. The output is the new Y calculation. The function, in its simplest form, can have the following C definition:

bool LOESS(const double* X, const double* Y, double *NewY,  int N, double SMOOTH_FACTOR); 

In SQL the equivalent used for updates or select should look like:

update my_table set new_y=loess(x_data, y_data, smooth_factor);

Obviously we don't need N and we may easily use the out parameter as the returning value. 

To calculate NEW_Y, the full vector X_DATA and Y_DATA are required in full upfront, and the resulting vector needs to be properly aligned (NEW_Y[i] needs to be paired with X[i]). Therefore this is not a scalar function, nor a table function. IMO this qualifies as a vector function, because it operates over vectors of common length (the fact some parameters are constant, does not affect the logic).

R works with vectors, C++ Eigen works with vectors, Pandas works with vectors, ... and SQL obviously works with vectors. I think that extending SQL using vector functions as described is quite realistic.

I hope this clarifies my previous post.

**Reply by Tishj** [reply]
If it needs all the data up front, I'd argue this is an aggregate function, no?

**Comment by danz68** [comment]
Not really, no. SQL-wise, an aggregation means reducing data to a meaningful scalar value, usually a number (but not only). Full data is required because every coordinate has an influence on the resulting set. For the given example, the output (new Y) is a vector of the same size as the input X or Y. It doesn't reduce it, so is not an aggregation. It doesn't make it scalar either, because we cant calculate any of the new coordinates independently. It is a vector function with a similar conceptual behavior as any math operator vec1 op vec2 = vec3. 

Imagine a multiple linear regression model, ok? What is the input? a few explanatory variables/vectors Xi and the response variable Y. The result is our model estimated new Y, which needs to be aligned with the existing data (every Y response aligned with its counterpart estimated_y). How do we write this in SQL? 

update our_model_dataset model_estimated_y = mline_regression(X1, X2, ..., Xn, Y) where ...;

I see no reasonable way of doing it using aggregates.

**Reply by Tishj** [reply]
Ah okay, so when given X values, it still outputs X values (scalar) but the X-1th input value could have impact on the first output value?

I think that would currently require a custom physical operator, so definitely something that an extension could achieve

**Reply by danz68** [reply]
Still not there, I guess.

Have been given A[0..n], B[0..n] as vectors, a vector function will return a vector of the same size Z[0..n]. Two aspects:

1. Internally, every Z[i] is calculated using all values of the vectors A, B (2n values), much more than what a scalar would use (just 2). 

2. The order needs to be considered relevant as well. Say, 
for A= {1,2,3} and B={2,3,5}, 
the output is Z={11,13,17}. 
But for a permutation, say A= {2,1,3} and B={3,2,5}, the output might be Z={14, 12, 19}. The result should not be considered just a permutation of the previous values.

I don't know much about DuckDB internals, if this requires a physical operator or so. But I can certainly say that functions using equal sized vectors as parameters and results are used everywhere. And, language-wise this is the natural way of looking at SQL queries, too.

**Reply by Tishj** [reply]
I think we're going in circles, let me clarify what I said earlier:

when I say "values", I mean tuples, rows.
Because of this part I thought you were talking about an aggregate function.
> To calculate NEW_Y, the full vector X_DATA and Y_DATA are required in full upfront, and the resulting vector needs to be properly aligned (NEW_Y[i] needs to be paired with X[i]). Therefore this is not a scalar function, nor a table function.

In my mind, scalar means every tuple produces a single output, hence:
> so when given X values, it still outputs X values (scalar)

But I think it's now clear to me what you're saying and I stand by what I said, this would require a special operator the way you're thinking of it.

It's likely possible to make a scalar out of this when you change the parameters to use lists instead, but one would need to verify at runtime if the provided list lengths are correct which is not very efficient.
There is work being done on adding fixed-size lists which would alleviate this issue however

**Comment by danz68** [comment]
Thank you for taking the time to chat about this, @Tishj. 
If there are only semantics differences, is fine.

IMO the UDF API feature names I suggested (ScalarUDF, AggregateUDF and VectorUDF) are closer to the development knowledge a data science coder has. Disregarding the language, one usually writes functions invoking scalar or vector variables (parameters or returns). 

Other than that, I would rather suggest to avoid fully materializing arrays/vectors (as lists or arrow objects) when passing data from duckdb via UDF APl. Memory-wise, a streaming chunks technique would be the safest bet.
